crates/sql_provider_datafusion/src/dbconnection.rs

```diff
@@ -1,20 +1,22 @@
 use std::any::Any;
 
-use ::duckdb::arrow::array::RecordBatch;
-use datafusion::{arrow::datatypes::SchemaRef, sql::TableReference};
+use datafusion::{
+    arrow::datatypes::SchemaRef, execution::SendableRecordBatchStream, sql::TableReference,
+};
 
 pub mod duckdbconn;
+pub mod postgresconn;
 
 pub type Error = Box<dyn std::error::Error + Send + Sync>;
 type Result<T, E = Error> = std::result::Result<T, E>;
 
-pub trait DbConnection<T, P> {
+pub trait DbConnection<T, P>: Send {
     fn new(conn: T) -> Self
     where
         Self: Sized;
-    fn get_schema(&mut self, table_reference: &TableReference) -> Result<SchemaRef>;
-    fn query_arrow(&mut self, sql: &str, params: &[P]) -> Result<Vec<RecordBatch>>;
-    fn execute(&mut self, sql: &str, params: &[P]) -> Result<u64>;
+    fn get_schema(&self, table_reference: &TableReference) -> Result<SchemaRef>;
+    fn query_arrow(&self, sql: &str, params: &[P]) -> Result<SendableRecordBatchStream>;
```

Comment:
Can we make these async functions so that we don't need to block the tokio executor?

Reply:
Unfortunately as we've currently structured our code it's not that simple because the DuckDB connection and pool library we're using is not thread safe. Yong is looking into the right thing to do here: https://github.com/spicehq/spiceai/issues/151